libgalois/include/katana/PropertyGraph.h

```diff
@@ -473,80 +487,200 @@ class KATANA_EXPORT PropertyGraph {
+    if (set_of_names.size() == 1) {
+      return *set_of_names.cbegin();
+    }
+    return std::nullopt;
   }
 
-  /// \returns the number of node TypeSetIDs (including kUnknownType)
-  size_t GetNodeTypeSetIDsNum() const {
-    return node_type_set_id_to_type_names_.size();
+  /// \returns the set of node EntityTypeIDs that contain
+  /// the node type with @param name
+  /// (assumes that the node type exists)
+  const SetOfEntityTypeIDs& NodeTypeNameToEntityTypeIDs(
+      const std::string& name) const {
+    return node_type_name_to_entity_type_ids_.at(name);
   }
 
-  /// \returns the number of edge TypeSetIDs (including kUnknownType)
-  size_t GetEdgeTypeSetIDsNum() const {
-    return edge_type_set_id_to_type_names_.size();
+  /// \returns the set of edge EntityTypeIDs that contain
+  /// the edge type with @param name
+  /// (assumes that the edge type exists)
+  const SetOfEntityTypeIDs& EdgeTypeNameToEntityTypeIDs(
+      const std::string& name) const {
+    return edge_type_name_to_entity_type_ids_.at(name);
+  }
+
+  /// \returns the set of atomic node subtype names that intersect
+  /// the node EntityTypeID @param node_entity_type_id
+  /// (assumes that the node EntityTypeID exists)
+  const SetOfTypeNames& NodeEntityTypeIDToAtomicSubtypeNames(
+      EntityTypeID node_entity_type_id) const {
+    return node_entity_type_id_to_type_names_.at(node_entity_type_id);
+  }
+
+  /// \returns the set of atomic edge subtype names that intersect
+  /// the edge EntityTypeID @param edge_entity_type_id
+  /// (assumes that the edge EntityTypeID exists)
+  const SetOfTypeNames& EdgeEntityTypeIDToAtomicSubtypeNames(
+      EntityTypeID edge_entity_type_id) const {
+    return edge_entity_type_id_to_type_names_.at(edge_entity_type_id);
+  }
+
+  /// \returns the set of atomic node subtype names that intersect
+  /// the node EntityTypeID @param node_entity_type_id
+  /// (assumes that the node EntityTypeID exists)
+  TypeNamesRange NodeEntityTypeIDToAtomicSubtypeNamesRange(
+      EntityTypeID node_entity_type_id) const {
+    auto& type_names =
+        NodeEntityTypeIDToAtomicSubtypeNames(node_entity_type_id);
+    ;
+    return MakeStandardRange(type_names.cbegin(), type_names.cend());
+  }
+
+  /// \returns the set of atomic edge subtype names that intersect
+  /// the edge EntityTypeID @param edge_entity_type_id
+  /// (assumes that the edge EntityTypeID exists)
+  TypeNamesRange EdgeEntityTypeIDToAtomicSubtypeNamesRange(
+      EntityTypeID edge_entity_type_id) const {
+    auto& type_names =
+        EdgeEntityTypeIDToAtomicSubtypeNames(edge_entity_type_id);
+    ;
+    return MakeStandardRange(type_names.cbegin(), type_names.cend());
+  }
+
+  /// \returns true iff the node type @param sub_type is a
+  /// sub-type of the node type @param super_type
+  /// (assumes that the sub_type and super_type EntityTypeIDs exists)
+  bool IsNodeSubtypeOf(EntityTypeID sub_type, EntityTypeID super_type) const {
```

Comment:
This is a bad implementation. We're checking for strings from one set in another. This is where we should be keeping maps to bit-sets (instead of strings) and compute the result much cheaper.